uefi/src/proto/shell_params.rs

```diff
@@ -0,0 +1,49 @@
+//! `ShellParams` protocol
+
+use crate::proto::unsafe_protocol;
+use crate::Char16;
+use core::ffi::c_void;
+
+#[cfg(feature = "alloc")]
+use crate::CStr16;
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+#[cfg(feature = "alloc")]
+use alloc::string::ToString;
+#[cfg(feature = "alloc")]
+use alloc::vec;
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+
+type ShellFileHandle = *const c_void;
+
+/// The ShellParameters protocol.
+#[repr(C)]
+#[unsafe_protocol("752f3136-4e16-4fdc-a22a-e5f46812f4ca")]
+pub struct ShellParameters {
+    /// Pointer to a list of arguments
+    pub argv: *const *const Char16,
+    /// Number of arguments
+    pub argc: usize,
+    /// Handle of the standard input
+    std_in: ShellFileHandle,
+    /// Handle of the standard output
+    std_out: ShellFileHandle,
+    /// Handle of the standard error output
+    std_err: ShellFileHandle,
+}
+
+impl ShellParameters {
+    /// Get a Vec of the shell parameter arguments
+    #[cfg(feature = "alloc")]
+    #[must_use]
+    pub fn get_args(&self) -> Vec<String> {
```

Comment:
Rather than collecting into a `Vec` here, could we return an iterator? That would allow it to work without `alloc`, and if the user actually needs a `Vec` they can call `.collect()` on it.

Reply:
Oh that's a good idea!

Do you think it would be useful to add something like this as well?

```rust
    /// Get a slice of the args, as Char16 pointers
    pub fn get_args_slice(&self) -> &[*const Char16] {
        unsafe { from_raw_parts(self.argv, self.argc) }
    }
```

Reply:
I'd suggest not exposing that as `pub fn` unless we have a specific real-world need for the API. Most of the uefi-rs API tries to provide safe wrappers rather than exposing raw pointers.

Reply:
What about returning a slice or an iterator to `CStr16`?
I couldn't figure out how to return a `&[CStr16]`, not sure if that's possible.
But an `Iterator<Item = CStr16>` would be doable. Or do you generally just want to return utf8 `String`s only?

Reply:
> I couldn't figure out how to return a &[CStr16], not sure if that's possible.

Does this help?
- https://docs.rs/uefi/0.20.0/uefi/data_types/struct.CStr16.html#method.from_ptr
- https://docs.rs/uefi/0.20.0/uefi/data_types/struct.CStr16.html#method.as_slice_with_nul

Reply:
Hmm but can I turn a `&[*const Char16]` into a `&[CStr16]`?
Not without copying, right?